flink-streaming-connectors/flink-connector-kafka-base/src/test/java/org/apache/flink/streaming/connectors/kafka/AtLeastOnceProducerTest.java

```diff
@@ -0,0 +1,206 @@
+import org.apache.flink.configuration.Configuration;
+import org.apache.flink.streaming.connectors.kafka.testutils.MockRuntimeContext;
+import org.apache.flink.streaming.util.serialization.KeyedSerializationSchema;
+import org.apache.flink.streaming.util.serialization.KeyedSerializationSchemaWrapper;
+import org.apache.flink.streaming.util.serialization.SimpleStringSchema;
+import org.apache.kafka.clients.producer.Callback;
+import org.apache.kafka.clients.producer.KafkaProducer;
+import org.apache.kafka.clients.producer.Producer;
+import org.apache.kafka.clients.producer.ProducerRecord;
+import org.apache.kafka.clients.producer.RecordMetadata;
+import org.apache.kafka.common.Metric;
+import org.apache.kafka.common.MetricName;
+import org.apache.kafka.common.PartitionInfo;
+import org.apache.kafka.common.serialization.ByteArraySerializer;
+import org.junit.Assert;
+import org.junit.Test;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.Properties;
+import java.util.concurrent.Future;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+/**
+ * Test ensuring that the producer is not dropping buffered records
+ */
+@SuppressWarnings("unchecked")
+public class AtLeastOnceProducerTest {
+
+	@Test
+	public void testAtLeastOnceProducer() throws Exception {
+		runTest(true);
+	}
+
+	// This test ensures that the actual test fails if the flushing is disabled
+	@Test(expected = AssertionError.class)
+	public void ensureTestFails() throws Exception {
+		runTest(false);
+	}
+
+	private void runTest(boolean flushOnCheckpoint) throws Exception {
+		Properties props = new Properties();
+		final TestingKafkaProducer<String> producer = new TestingKafkaProducer<>("someTopic", new KeyedSerializationSchemaWrapper<>(new SimpleStringSchema()), props);
+		producer.setFlushOnCheckpoint(flushOnCheckpoint);
+		producer.setRuntimeContext(new MockRuntimeContext(0, 1));
+
+		producer.open(new Configuration());
+
+		for(int i = 0; i < 100; i++) {
+			producer.invoke("msg-" + i);
+		}
+		// start a thread confirming all pending records
+		final Tuple1<Throwable> runnableError = new Tuple1<>(null);
+		final AtomicBoolean markOne = new AtomicBoolean(false);
+		Runnable confirmer = new Runnable() {
+			@Override
+			public void run() {
+				try {
+					MockProducer mp = producer.getProducerInstance();
+					List<Callback> pending = mp.getPending();
+
+					// we ensure thread A is locked and didn't reach markOne
+					// give thread A some time to really reach the snapshot state
+					Thread.sleep(500);
+					if(markOne.get()) {
+						Assert.fail("Snapshot was confirmed even though messages " +
+								"were still in the buffer");
+					}
+					Assert.assertEquals(100, pending.size());
+
+					// now confirm all checkpoints
+					for(Callback c: pending) {
+						c.onCompletion(null, null);
+					}
+					pending.clear();
+					// wait for the snapshotState() method to return
+					Thread.sleep(100);
```

Comment:
Here again, who tells us that Thread A has progressed after 100 ms to the point where he set `markOne` to true.